feat(acp): Add ACP protocol support for OpenCode integration

oliveiraantoniocc opened the pull request:
## Summary

This PR adds support for the Agent Client Protocol (ACP) to enable native integration with OpenCode and other ACP-compatible agents.

## Key Changes

### ACP Protocol Implementation
- **ACP Client** (`acp-client.ts`): JSON-RPC 2.0 client for stdio communication with ACP agents
- **ACP Process** (`acp-process.ts`): Wrapper that provides ProcessManager-compatible interface for ACP agents
- **ACP Adapter** (`acp-adapter.ts`): Converts ACP session updates to Maestro's ParsedEvent format
- **ACP Types** (`types.ts`): Complete TypeScript definitions for ACP protocol spec

### Bug Fixes
- **Historical Message Accumulation**: Fixed issue where loading existing ACP sessions would replay all historical messages and accumulate them in the UI. Added `isLoadingSession` flag to ignore session updates during `session/load`.
- **Usage Stats**: Added support for forwarding token usage statistics from ACP responses (though OpenCode currently doesn't provide this data per the optional ACP spec).

### Configuration & Debugging
- **Agent Configuration**: Added `useACP` and `acpShowStreaming` options to agent configs for per-agent ACP mode control
- **Debug Logging**: Added `MAESTRO_LOG_LEVEL` environment variable support and `dev:main:debug` script for easier debugging
- **Spawn Logic**: Updated process spawning to check agent capabilities and config to determine ACP vs CLI mode

### OpenCode Integration
- **Detection**: OpenCode is now auto-detected and configured with ACP support
- **Capabilities**: Marked OpenCode as supporting ACP, session resume, images, and JSON output
- **Session Management**: Load/resume OpenCode sessions via ACP `session/load` method

## Testing

Tested with OpenCode v1.0.190:
- ✅ New sessions create and stream responses correctly
- ✅ Session resume loads history without accumulation
- ✅ Streaming text displays properly in UI
- ✅ Tool calls and thinking blocks are parsed
- ✅ Historical messages are filtered during session load

## Known Limitations

1. **Token Usage**: OpenCode doesn't currently report token usage in ACP responses (usage field is optional per spec)
2. **Session Cleanup**: Agent sessions remain in agent storage after Maestro session deletion (needs ACP session deletion support)

## Checklist

- [x] Code follows project style guidelines
- [x] Tested locally with OpenCode
- [x] Commit messages follow conventional commits
- [x] Fixed historical message accumulation bug
- [x] Added debug logging improvements

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **New Features**
  * Full ACP support: interactive streaming agent sessions, resumable sessions, image-enabled prompts, richer session events (results, errors, tool usage) and session controls (cancel/interrupt).
  * ACP-backed process path and spawn mode for agents, with option to resume via session ID.
  * Agent settings expose ACP options (including streaming toggle) marked as Beta.

* **Documentation**
  * Release notes formatting normalized for clearer rendering.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

pedramamini commented:
@oliveiraantoniocc can i trouble you to rebase this on latest refactored main? because it's feature-gated, i think we can just merge this in from there and get feedback from those who are using it. can you ensure it has a `BETA` pill too? like i do with this setting here:

<img width="618" height="287" alt="image" src="https://github.com/user-attachments/assets/2994a4f3-4f57-4c79-b9ff-e01c65976afc" />

stefandevo commented:
Would love to see this!

oliveiraantoniocc committed:
feat: add ACP (Agent Client Protocol) support for OpenCode

- Add ACP module (types, client, process, adapter) for JSON-RPC communication
- Add supportsACP capability flag (only OpenCode has true)
- Add useACP and acpShowStreaming config options for OpenCode
- Create AcpSpawner in process-manager/spawners/ for ACP mode spawning
- Update ProcessManager with spawnAsync() method for ACP initialization
- Update IPC handler to detect ACP mode and use async spawning
- Add BETA badge for ACP options in AgentConfigPanel
- Update capability tests to include supportsACP

ACP enables structured JSON-RPC communication with OpenCode instead of PTY,
providing cleaner message parsing and better error handling.

oliveiraantoniocc committed:
style: format ACP files with Prettier

oliveiraantoniocc committed:
fix: address CodeRabbit review feedback

- Add dynamic version via getAppVersion() utility (replaces hardcoded '0.12.0')
- Add ContentBlockFlat type for OpenCode's flat format compatibility
- Improve type documentation for ContentBlock formats
- Fix docs/releases.md Prettier formatting (pre-existing upstream issue)

The following issues were already addressed in previous commits:
- SSH wrapping and customArgs support (ACPProcessConfig.args)
- Terminal capability disabled until handlers implemented
- Images only sent on initial prompt (imagesConsumed flag)
- Streaming dedup fix (cross-prompt overlap handling)
- Log redaction for ACP payloads (redactForLogging)
- Sentry integration for unexpected errors (captureError)
- fs:read/fs:write proper error responses

oliveiraantoniocc committed:
fix: wait for state update between keyboard navigation events in test

The ArrowUp navigation test was flaky because it fired two keydown events
synchronously without waiting for React to process the state update from
the first event. Now we properly await each state transition before
proceeding to the next keyboard action.

oliveiraantoniocc committed:
test: add comprehensive ACP module tests

Add unit tests for the ACP module covering:
- Type definitions and protocol constants (acp-types.test.ts)
- SessionUpdate to ParsedEvent adapter functions (acp-adapter.test.ts)
- ACPClient class initialization and event emitter (acp-client.test.ts)

47 tests total covering core ACP functionality.

oliveiraantoniocc committed:
feat(acp): add usage_update and config_option_update types

Based on OpenCode ACP implementation analysis:
- Add UsageUpdate type for token usage and cost tracking (used, size, cost)
- Add ConfigOptionUpdate type for session configuration changes
- Update acp-adapter.ts to handle both new update types
- Convert current_mode_update to system message (was returning null)
- Add comprehensive tests for new types (16 new tests)

The usage_update maps to ParsedEvent 'usage' type with:
- used -> inputTokens (cumulative context usage)
- size -> contextWindow
- cost.amount -> costUsd (when currency is USD)

The config_option_update maps to 'system' type with formatted message.

oliveiraantoniocc committed:
feat(acp): add ACP error type detection for proper error categorization

Replace hardcoded 'unknown' error types in ACPProcess with intelligent
error detection via new acp-error-detector module. Errors are now
categorized by JSON-RPC error codes, OpenCode error patterns, and
keyword-based fallback matching into auth_expired, rate_limited,
token_exhaustion, network_error, session_not_found, permission_denied,
and agent_crashed types for better user feedback and recovery actions.

oliveiraantoniocc committed:
feat(acp): add Gemini CLI ACP support with configurable args, authentication, and error patterns

- Add acpArgs field to AgentConfig/ProcessConfig for per-agent ACP arguments
  (OpenCode: ['acp'], Gemini CLI: ['--acp'])
- Add authenticate() method to ACPClient for Gemini CLI's required auth flow
- Add useFlatContentBlocks config flag for OpenCode vs spec content format
- Enable supportsACP: true for gemini-cli agent with useACP/acpShowStreaming options
- Add Gemini CLI error patterns (Google OAuth, API key, RESOURCE_EXHAUSTED,
  PERMISSION_DENIED, UNAVAILABLE, login required)
- Update acp-error-detector to support multiple agent types with agentType param
- Add 12 new tests for Gemini CLI patterns and agentType prioritization
- All 111 ACP tests passing

oliveiraantoniocc commented:
@coderabbitai review
